### PR TITLE
check for overflow and underflow while choosing proposer

### PR DESCRIPTION
**Pull request by melekes**
Refs #919

### CODE REVIEW

File: types/validator_set.go [hunk 1]
```diff
@@ -48,22 +48,34 @@ func NewValidatorSet(vals []*Validator) *ValidatorSet {
...
+	totalVotingPower := valSet.TotalVotingPower()
 	for i := 0; i < times; i++ {
 		mostest := validatorsHeap.Peek().(*Validator)
 		if i == times-1 {
 			valSet.Proposer = mostest
 		}
-		mostest.Accum -= int64(valSet.TotalVotingPower())
+
+		// mind underflow
+		if mostest.Accum >= mostNegative+totalVotingPower {
```
Comment (melekes) [inline types/validator_set.go]: write a test for it

File: types/validator_set.go [hunk 2]
```diff
@@ -117,7 +129,13 @@ func (valSet *ValidatorSet) Size() int {
 func (valSet *ValidatorSet) TotalVotingPower() int64 {
 	if valSet.totalVotingPower == 0 {
 		for _, val := range valSet.Validators {
-			valSet.totalVotingPower += val.VotingPower
+			// mind overflow
+			if valSet.totalVotingPower <= mostPositive-val.VotingPower {
```
Comment (melekes) [inline types/validator_set.go]: write a test for it

**Comment by melekes** [comment]
1) Should I write a comment that the reason for these checks is that it is critical code?
2) @jaekwon is this what you had in mind by "mind the overflow". What's the word mind mean? Be mindful or actually add the checks plus tests like a did?

File: types/validator_set.go [hunk 3]
```diff
@@ -48,12 +48,17 @@ func NewValidatorSet(vals []*Validator) *ValidatorSet {
...
 // incrementAccum and update the proposer
-// TODO: mind the overflow when times and votingPower shares too large.
 func (valSet *ValidatorSet) IncrementAccum(times int) {
 	// Add VotingPower * times to each validator and order into heap.
 	validatorsHeap := cmn.NewHeap()
 	for _, val := range valSet.Validators {
-		val.Accum += val.VotingPower * int64(times) // TODO: mind overflow
+		res, overflow := signedMulWithOverflowCheck(val.VotingPower, int64(times))
+		// check for overflow both multiplication and sum
+		if !overflow && val.Accum <= mostPositive-res {
```
Comment (jaekwon) [inline types/validator_set.go]: lets make addSafe().

File: types/validator_set.go [hunk 4]
```diff
@@ -48,12 +48,17 @@ func NewValidatorSet(vals []*Validator) *ValidatorSet {
 }
 
 // incrementAccum and update the proposer
-// TODO: mind the overflow when times and votingPower shares too large.
 func (valSet *ValidatorSet) IncrementAccum(times int) {
 	// Add VotingPower * times to each validator and order into heap.
 	validatorsHeap := cmn.NewHeap()
 	for _, val := range valSet.Validators {
-		val.Accum += val.VotingPower * int64(times) // TODO: mind overflow
+		res, overflow := signedMulWithOverflowCheck(val.VotingPower, int64(times))
```
Comment (jaekwon) [inline types/validator_set.go]: s/signedMulWithOverflowCheck/mulSafe/g
I don't think we'd check overflow with unsigned numbers.

Comment (jaekwon) [reply types/validator_set.go]: or multiply unsigned numbers for that matter.  if we do it's probably for bit-twiddling and we'll prob want to create custom functions for those kinds of operations, I think.

File: types/validator_set.go [hunk 5]
```diff
@@ -63,7 +68,13 @@ func (valSet *ValidatorSet) IncrementAccum(times int) {
 		if i == times-1 {
 			valSet.Proposer = mostest
 		}
-		mostest.Accum -= int64(valSet.TotalVotingPower())
+
+		// mind underflow
```
Comment (jaekwon) [inline types/validator_set.go]: I'd add underflow protection to addSafe() as well.  maybe we can have subSafe() too, which handles both over/underflow protection.

File: types/validator_set.go [hunk 6]
```diff
@@ -117,7 +128,13 @@ func (valSet *ValidatorSet) Size() int {
 func (valSet *ValidatorSet) TotalVotingPower() int64 {
 	if valSet.totalVotingPower == 0 {
 		for _, val := range valSet.Validators {
-			valSet.totalVotingPower += val.VotingPower
+			// mind overflow
```
Comment (jaekwon) [inline types/validator_set.go]: ditto

File: types/validator_set.go [hunk 7]
```diff
@@ -48,12 +48,22 @@ func NewValidatorSet(vals []*Validator) *ValidatorSet {
...
 
 // incrementAccum and update the proposer
-// TODO: mind the overflow when times and votingPower shares too large.
 func (valSet *ValidatorSet) IncrementAccum(times int) {
 	// Add VotingPower * times to each validator and order into heap.
 	validatorsHeap := cmn.NewHeap()
 	for _, val := range valSet.Validators {
-		val.Accum += val.VotingPower * int64(times) // TODO: mind overflow
+		// check for overflow both multiplication and sum
+		res, overflow := safeMul(val.VotingPower, int64(times))
```
Comment (melekes) [inline types/validator_set.go]: maybe we can simplify this somehow?

Comment (jaekwon) [reply types/validator_set.go]: safeMulClip could automatically set to math.MaxInt64 or math.MinInt64 appropriately, and return a single value.  safeAddClip likewise.

File: types/validator_set.go [hunk 8]
```diff
@@ -425,3 +449,39 @@ func RandValidatorSet(numValidators int, votingPower int64) (*ValidatorSet, []*P
 	sort.Sort(PrivValidatorsByAddress(privValidators))
 	return valSet, privValidators
 }
+
+///////////////////////////////////////////////////////////////////////////////
+// Safe multiplication and addition/subtraction
+
+const mostNegative int64 = -mostPositive - 1
```
Comment (jaekwon) [inline types/validator_set.go]: math.MaxInt64 and math.MinInt64

**Comment by ebuchman** [comment]
Is this really how we should deal with the overflows - ie. just use the max ints? Or should we consider   big int arithmetic ?

**Pull request merged**